src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Unit.Tests/CompilationHelper.cs

```diff
@@ -275,6 +275,36 @@ public class Location
             return CreateCompilation(source);
         }
 
+        public static Compilation CreateCompilationWithRecordPositionalParameters()
```

Comment:
Could you also add a test for regular class like this one?
```C#
public class MyClass
{
    public MyClass(int value)
    {
        Value = value;
    }

    public int Value { get; init; }
}
```
This one is also mapping the `Value` property to the constructor parameter.

Reply:
Yes, that's true ... but that's exactly the same code that will be generated from the equivalent record type `public record MyClass(int Value)`. Given that, do you still feel the additional test is valuable? If so, I'm happy to add it.

Reply:
We just need to make sure that the parameter-to-property matching logic is compatible with what the runtime serializer is doing. I believe your changes make a case sensitive comparison so I think it leaves out a few false positives.

Reply:
Note the difference in casing between the parameter `value` and the property `Value`.  I believe the matching logic is case insensitive and doesn't normalize to camelCase, but a few more tests should validate this.

Reply:
Might also need to validate that this class does emit warnings:
```C#
public class MyClass
{
    public MyClass(int value)
    {
        Value = value;
    }

    public int Value { get; init; }
    public string Orphaned { get; init; }
}
```

Reply:
I was just mulling over the casing mismatch issue :)

Yes, the `PropertyIsConstructorParameter` check is case sensitive, so it _would_ emit the SYSLIB1037 warning on that `Value` property.

The `Orphaned` property should definitely emit the warning.

I'll get the new test cases added!

Reply:
My intent is to code the test for `MyClass` to assert that it _does_ emit a warning for the case-mismatched `value` parameter and `Value` property. I'm not totally sure how the source generator handles that and what the runtime behavior would be, so for this PR I'm more comfortable being conservative and just squashing the warnings that I know for certain are false positives, since that's enough to get a baseline record type to a good place without accidentally breaking anything else in the process.

Are you on board with that? Or are you saying that we should address this, so that the first version of `MyClass` without the `Orphaned` property _doesn't_ emit the warning?

Reply:
I'm seeing a couple of possible alternatives:

* Either restrict your change to positional records, regular classes keep getting the same warning regardless of shape.
* Address the issue fully and make the parameter matching case insentive.

Either seems fine to me.

Reply:
Hmm... We can test whether a System.Type pertains to a record by checking for the presence of a method named `<Clone>$`, but that's taking a dependency on an implementation detail of the compiler. That's probably not a great thing to be doing if it can be avoided, so I'll go with the case insensitive name match.

Reply:
I would guess that Roslyn has metadata indicating whether a type symbol is a positional record, so such a heuristic might not be necessary. I could be wrong though.

Reply:
Yeah, given Roslyn's parse tree, we could definitely find that out. But if we went that route, then we end up with different rules based on whether the record type and the serialization context are in the same project or not - a record type from a referenced library would be treated like a plain class. I'd **very** much like to avoid that.

Reply:
If it's not possible to determine whether a type in a referenced library is a positional record, I agree.

Reply:
@eiriktsarpalis The parameter matching has been updated to be case insensitive, and the two additional tests have been added.

Reply:
Changes look good, thanks! It seems that our acceptance test suite doesn't have any coverage for positional records. Would it be possible to add a simple test validating that sourcegen deserialization works as expected? (and that therefore the change in the diagnostic is a sound one).

To help you out, that would involve:

1. Adding a positional record to the [TestClasses.cs](https://github.com/dotnet/runtime/blob/a62c435d1d65d16af1337eecf66f6e753caefdf7/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/TestClasses.cs) file.
2. Updating the [ITestContext](https://github.com/dotnet/runtime/blob/a62c435d1d65d16af1337eecf66f6e753caefdf7/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/ContextClasses.cs#L11) interface to include an entry for the positional record.
3. Update all `JsonSerializerContext` annotations implementing `ITestContext`.
4. Add a unit test for serialization and deserialization in the [`RealWorldContextTests`](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Tests/RealWorldContextTests.cs) class.

Reply:
Sure, and done!